relengapi/blueprints/slaveloan/tasks.py

```diff
@@ -0,0 +1,199 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import random
+import socket
+
+from furl import furl
+
+import bzrest
+import requests
+
+from requests import RequestException
+
+from flask import current_app
+from functools import wraps
+from redo import retry
+from relengapi.blueprints.slaveloan import bugzilla
+from relengapi.blueprints.slaveloan import slave_mappings
+from relengapi.blueprints.slaveloan.model import History
+from relengapi.blueprints.slaveloan.model import Loans
+from relengapi.blueprints.slaveloan.model import Machines
+from relengapi.lib.celery import task
+from relengapi.util import tz
+
+import celery
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def add_task_to_history(loanid, msg):
+    session = current_app.db.session('relengapi')
+    l = session.query(Loans).get(loanid)
+    history = History(for_loan=l,
+                      timestamp=tz.utcnow(),
+                      msg=msg)
+    session.add(history)
+    session.commit()
+    logger.debug("Log_line: %s" % msg)
+
+
+def add_to_history(before=None, after=None):
+    def decorator(f):
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            bound_task = None
+            loanid = kwargs.get("loanid", None)
+            if args and isinstance(args[0], celery.Task):
+                bound_task = args[0]
+            if before:
+                add_task_to_history(loanid, before.format(**locals()))
+            retval = f(*args, **kwargs)
+            if after:
+                add_task_to_history(loanid, after.format(**locals()))
+            return retval
+        return wrapper
+    return decorator
+
+
+@task(bind=True)
+@add_to_history(
+    before="Choosing an inhouse machine based on slavealloc",
+    after="Chose inhouse machine {retval!s}")
+def choose_inhouse_machine(self, loanid, loan_class):
+    logger.debug("Choosing inhouse machine")
+    url = furl(current_app.config.get("SLAVEALLOC_URL", None))
+    # XXX: ToDo raise fatal if no slavealloc
+    url.path.add("slaves")
+    url.args["enabled"] = 1
+    try:
+        all_slaves = requests.get(str(url)).json()
+    except RequestException as exc:
+        logger.exception("Exception: %s" % exc)
+        self.retry(exc=exc)
+    # pylint silence
+    # available_slaves = filter(slave_mappings.slave_filter(loan_class), all_slaves)
+    available_slaves = [slave for slave in all_slaves
+                        if slave_mappings.slave_filter(loan_class)(slave)]
+    chosen = random.choice(available_slaves)
+    logger.debug("Chosen Slave = %s" % chosen)
+    return chosen['name']
+
+
+@task(bind=True, max_retries=None)
+@add_to_history(
+    before="Identifying FQDN and IP of {args[1]}",
+    after="Aquired FQDN and IP")
+def fixup_machine(self, machine, loanid):
```

Comment:
No, not meant to be used with ec2 (we don't currently set CNAME's for ec2 hosts), however, its seperate because of my desire to make every celery task idempotent.